CMakeLists.txt

```diff
@@ -0,0 +1,857 @@
+
+if(MINGW)
+  check_symbol_exists("sleep" "unistd.h" HAVE_DECL_SLEEP)
+  check_symbol_exists("nanosleep" "time.h" HAVE_DECL_NANOSLEEP)
+endif()
+
+if(LINUX)
+  check_c_source_compiles("
+    #include <signal.h>
+    #include <time.h>
+    int main() { return SIGEV_THREAD_ID || CLOCK_THREAD_CPUTIME_ID; }"
+    HAVE_LINUX_SIGEV_THREAD_ID)
+endif()
+
+configure_file(cmake/config.h.in ${CMAKE_BINARY_DIR}/config.h @ONLY)
+configure_file(cmake/tcmalloc.h.in
+               ${CMAKE_BINARY_DIR}/gperftools/tcmalloc.h
+               @ONLY)
+
+if(gperftools_build_cpu_profiler OR
+   gperftools_build_heap_profiler OR
+   gperftools_build_heap_checker)
+  set(with_stack_trace ON)
+endif()
+
+if(gperftools_enable_frame_pointers AND
+   NOT unwind_libs AND
+   NOT gperftools_build_minimal)
+  if(HAVE_UNWIND_BACKTRACE)
+    message(WARNING "No frame pointers and no libunwind. "
+                    "Using experimental backtrace capturing via libgcc. "
+                    "Expect crashy cpu profiler.")
+    if(gperftools_enable_stacktrace_via_backtrace)
+      message(WARNING "No frame pointers and no libunwind. "
+                      "Using experimental backtrace(). "
+                      "Expect crashy cpu profiler.")
+    else()
+      message(FATAL_ERROR "No frame pointers and no libunwind. "
+                          "The compilation will fail.")
+    endif()
+  endif()
+endif()
+
+# Based on Makefile.am
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+list(APPEND inc $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src>)
+if(NOT with_stack_trace)
+  list(APPEND defs NO_TCMALLOC_SAMPLES)
+endif()
+
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  list(APPEND flags
+    -Wall -Wwrite-strings -Woverloaded-virtual
+    -Wno-sign-compare
+    -fno-builtin-malloc -fno-builtin-free -fno-builtin-realloc
+    -fno-builtin-calloc -fno-builtin-cfree
+    -fno-builtin-memalign -fno-builtin-posix_memalign
+    -fno-builtin-valloc -fno-builtin-pvalloc)
+
+  if(i386)
+    list(APPEND flags -mmmx -fno-omit-frame-pointer)
+  endif()
+endif()
+
+if(have_w_no_unused_result)
+  list(APPEND flags -Wno-unused-result)
+endif()
+
+if(have_sized_deallocation)
+  list(APPEND flags -fsized-deallocation)
+endif()
+
+list(APPEND libs ${libstdcxx_la_linker_flag})
+
+if(gperftools_enable_frame_pointers)
+  list(APPEND flags -fno-omit-frame-pointer)
+elseif(NOT CMAKE_CXX_FLAGS MATCHES "-fno-omit-frame-pointer")
+  list(APPEND defs NO_FRAME_POINTER)
+endif()
+
+if(MINGW)
+  list(APPEND tcmalloc_lib_options "-Wl,-u__tcmalloc")
+endif()
+
+if(objcopy_exe)
+  set(weaken
+      -W malloc -W free -W realloc -W calloc -W cfree
+      -W memalign -W posix_memalign -W valloc -W pvalloc
+      -W aligned_alloc
+      -W malloc_stats -W mallopt -W mallinfo -W nallocx
+      -W _Znwm -W _ZnwmRKSt9nothrow_t -W _Znam -W _ZnamRKSt9nothrow_t
+      -W _ZdlPv -W _ZdaPv
+      -W __Znwm -W __ZnwmRKSt9nothrow_t -W __Znam -W __ZnamRKSt9nothrow_t
+      -W __ZdlPv -W __ZdaPv)
```

Comment:
This variable doesn't appear to be used anywhere. I was checking usage because this will initialize `weaken` with `-W:malloc:-W:free:-W:realloc` instead of spaces. Is this intentional? It's uncommon to use flags outside of double quotes.

Reply:
Double quotes are only necessary for interpolation and potentially ambiguous phrases. In CMake, pretty much everything defaults to a string. If you'd like, I can change it.

Reply:
That's not entirely true, though. `set()` with multiple arguments (beyond the initial name/type) create a list.

```cmake
set (FOO_BAR a b c d e f)
set (FOO_BAR2 "a b c d e f")
message (STATUS "${FOO_BAR}")
message (STATUS "${FOO_BAR2}")
```

outputs

```
-- a;b;c;d;e;f
-- a b c d e f
```

Reply:
Ya so this is one of the worse aspects of CMake. Lists are strings too, just delimited by semicolons. The problem with using a large string that contains all the flags is that COMMAND wants a list of args, not a string.

Reply:
Yeah :| As long as it's not used within a string interpolation expression, they'll work as expected.

I agree CMake misses the mark here.